src/ring_theory/discriminant.lean

```diff
@@ -0,0 +1,222 @@
+  (b : ι → B) := by { classical, exact (trace_matrix A b).det }
+
+lemma discr_def [decidable_eq ι] [fintype ι] (b : ι → B) :
+  discr A b = (trace_matrix A b).det := by convert rfl
+
+variable [fintype ι]
+
+section basic
+
+lemma discr_zero_of_not_linear_independent [is_domain A] {b : ι → B}
+  (hli : ¬linear_independent A b) : discr A b = 0 :=
+begin
+  classical,
+  obtain ⟨g, hg, i, hi⟩ := fintype.not_linear_independent_iff.1 hli,
+  have : (trace_matrix A b).mul_vec g = 0,
+  { ext i,
+    have : ∀ j, (trace A B) (b i * b j) * g j = (trace A B) (((g j) • (b j)) * b i),
+    { intro j, simp [mul_comm], },
+    simp only [mul_vec, dot_product, trace_matrix, pi.zero_apply, trace_form_apply,
+      λ j, this j, ← linear_map.map_sum, ← sum_mul, hg, zero_mul, linear_map.map_zero] },
+  by_contra h,
+  rw discr_def at h,
+  simpa [matrix.eq_zero_of_mul_vec_eq_zero h this] using hi,
+end
+
+variable {A}
+
+lemma discr_of_matrix_vec_mul [decidable_eq ι] (b : ι → B) (P : matrix ι ι A) :
+  discr A ((P.map (algebra_map A B)).vec_mul b) = P.det ^ 2 * discr A b :=
+by rw [discr_def, trace_matrix_of_matrix_vec_mul, det_mul, det_mul, det_transpose, mul_comm,
+    ← mul_assoc, discr_def, pow_two]
+
+
+lemma discr_of_matrix_mul_vec [decidable_eq ι] (b : ι → B) (P : matrix ι ι A) :
+  discr A ((P.map (algebra_map A B)).mul_vec b) = P.det ^ 2 * discr A b :=
+by rw [discr_def, trace_matrix_of_matrix_mul_vec, det_mul, det_mul, det_transpose,
+  mul_comm, ← mul_assoc, discr_def, pow_two]
+
+end basic
+
+section field
+
+variables (K : Type u) {L : Type v} (E : Type z) [field K] [field L] [field E]
+variables [algebra K L] [algebra K E] [algebra L E]
+variables [module.finite K L] [is_separable K L] [is_alg_closed E]
+variables (b : ι → L) (pb : power_basis K L)
+
+lemma discr_not_zero_of_linear_independent [nonempty ι]
+  (hcard : fintype.card ι = finrank K L) (hli : linear_independent K b) : discr K b ≠ 0 :=
+begin
+  classical,
+  have := span_eq_top_of_linear_independent_of_card_eq_finrank hli hcard,
+  rw [discr_def, trace_matrix_def],
+  simp_rw [← basis.mk_apply hli this],
+  rw [← trace_matrix_def, trace_matrix_of_basis, ← bilin_form.nondegenerate_iff_det_ne_zero],
+  exact trace_form_nondegenerate _ _
+end
+
+@[nolint unused_arguments]
+lemma _root_.algebra.trace_matrix_eq_embeddings_matrix_mul_trans :
+  (trace_matrix K b).map (algebra_map K E) =
+  (embeddings_matrix K E b) ⬝ (embeddings_matrix K E b)ᵀ :=
+by { ext i j, simp [trace_eq_sum_embeddings, embeddings_matrix, mul_apply] }
+
+lemma _root_.algebra.trace_matrix_eq_embeddings_matrix_reindex_mul_trans
+  (e : ι ≃ (L →ₐ[K] E)) : (trace_matrix K b).map (algebra_map K E) =
+  (embeddings_matrix_reindex K E b e) ⬝ (embeddings_matrix_reindex K E b e)ᵀ :=
+by rw [trace_matrix_eq_embeddings_matrix_mul_trans, embeddings_matrix_reindex, reindex_apply,
+  mul_transpose_eq_minor_mul_minor_transpose, ← equiv.coe_refl, equiv.refl_symm]
```

Comment:
Do you still need the `root`? Aren't you already in the `algebra` namespace now?